File: app/src/main/java/emu/skyline/LogActivity.java
package emu.skyline;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.preference.PreferenceManager;

import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.stream.Collectors;

import javax.net.ssl.HttpsURLConnection;

public class LogActivity extends AppCompatActivity {
    private File log_file;
    private LogAdapter adapter;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.log_activity);
        setSupportActionBar(findViewById(R.id.toolbar));
        final ActionBar actionBar = getSupportActionBar();
        if (actionBar != null)
            actionBar.setDisplayHomeAsUpEnabled(true);
        final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        ListView log_list = findViewById(R.id.log_list);
        adapter = new LogAdapter(this, prefs.getBoolean("log_compact", false), Integer.parseInt(prefs.getString("log_level", "3")), getResources().getStringArray(R.array.log_level));
        log_list.setAdapter(adapter);
        try {
            log_file = new File(getApplicationInfo().dataDir + "/skyline.log");
            final InputStream inputStream = new FileInputStream(log_file);
            final BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
            try {
                boolean done = false;
                while (!done) {
                    String line = reader.readLine();
                    if (!(done = (line == null))) {
                        adapter.add(line);
                    }
                }
            } catch (final IOException e) {
                Log.w("Logger", "IO Error during access of log file: " + e.getMessage());
                Toast.makeText(getApplicationContext(), getString(R.string.io_error) + ": " + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        } catch (final FileNotFoundException e) {
            Log.w("Logger", "IO Error during access of log file: " + e.getMessage());
            Toast.makeText(getApplicationContext(), getString(R.string.file_missing), Toast.LENGTH_LONG).show();
            finish();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.toolbar_log, menu);
        final MenuItem mSearch = menu.findItem(R.id.action_search_log);
        SearchView searchView = (SearchView) mSearch.getActionView();
        searchView.setSubmitButtonEnabled(false);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            public boolean onQueryTextSubmit(final String query) {
                searchView.setIconified(false);
                return false;
            }

            @Override
            public boolean onQueryTextChange(final String newText) {
                adapter.getFilter().filter(newText);
                return true;
            }
        });
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_clear:
                try {
                    final FileWriter fileWriter = new FileWriter(log_file, false);
                    fileWriter.close();
                } catch (final IOException e) {
                    Log.w("Logger", "IO Error while clearing the log file: " + e.getMessage());
                    Toast.makeText(getApplicationContext(), getString(R.string.io_error) + ": " + e.getMessage(), Toast.LENGTH_LONG).show();
                }
                Toast.makeText(getApplicationContext(), getString(R.string.cleared), Toast.LENGTH_LONG).show();
                finish();
                return true;
            case R.id.action_share_log:
                final Thread share_thread = new Thread(() -> {
                    HttpsURLConnection urlConnection = null;
                    try {
                        final URL url = new URL("https://hastebin.com/documents");
                        urlConnection = (HttpsURLConnection) url.openConnection();
                        urlConnection.setRequestMethod("POST");
                        urlConnection.setRequestProperty("Host", "hastebin.com");
                        urlConnection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
                        urlConnection.setRequestProperty("Referer", "https://hastebin.com/");
                        urlConnection.setRequestProperty("Connection", "keep-alive");
                        final OutputStream outputStream = new BufferedOutputStream(urlConnection.getOutputStream());
                        final BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
                        final FileReader fileReader = new FileReader(log_file);
                        int chr;
                        while ((chr = fileReader.read()) != -1) {
                            bufferedWriter.write(chr);
                        }
                        bufferedWriter.flush();
                        bufferedWriter.close();
                        outputStream.close();
                        if (urlConnection.getResponseCode() != 200) {
                            Log.e("LogUpload", "HTTPS Status Code: " + urlConnection.getResponseCode());
                            throw new Exception();
                        }
                        final InputStream inputStream = new BufferedInputStream(urlConnection.getInputStream());
                        final BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8));
                        final String key = new JSONObject(bufferedReader.lines().collect(Collectors.joining())).getString("key");
                        bufferedReader.close();
                        inputStream.close();
                        final String result = "https://hastebin.com/" + key;
                        final Intent sharingIntent = new Intent(Intent.ACTION_SEND).setType("text/plain").putExtra(Intent.EXTRA_TEXT, result);
                        startActivity(Intent.createChooser(sharingIntent, "Share log url with:"));
                    } catch (final Exception e) {
                        runOnUiThread(() -> Toast.makeText(getApplicationContext(), getString(R.string.share_error), Toast.LENGTH_LONG).show());
                        e.printStackTrace();
                    } finally {
                        assert urlConnection != null;
                        urlConnection.disconnect();
                    }
                });
                share_thread.start();
                try {
                    share_thread.join(1000);
                } catch (final Exception e) {
                    Toast.makeText(getApplicationContext(), getString(R.string.share_error), Toast.LENGTH_LONG).show();
                    e.printStackTrace();
                }
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/emu/skyline/LogActivity.kt
/*
 * SPDX-License-Identifier: MPL-2.0
 * Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
 */

package emu.skyline

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import emu.skyline.adapter.GenericAdapter
import emu.skyline.adapter.HeaderViewItem
import emu.skyline.adapter.LogViewItem
import emu.skyline.databinding.LogActivityBinding
import emu.skyline.utils.Settings
import org.json.JSONObject
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.net.URL
import javax.inject.Inject
import javax.net.ssl.HttpsURLConnection

@AndroidEntryPoint
class LogActivity : AppCompatActivity() {
    private val binding by lazy { LogActivityBinding.inflate(layoutInflater) }

    /**
     * The log file is used to read log entries from or to clear all entries
     */
    private lateinit var logFile : File

    private val adapter = GenericAdapter()

    @Inject
    lateinit var settings : Settings

    override fun onCreate(savedInstanceState : Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(binding.root)

        setSupportActionBar(binding.titlebar.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val compact = settings.logCompact
        val logLevel = settings.logLevel.toInt()
        val logLevels = resources.getStringArray(R.array.log_level)

        binding.logList.adapter = adapter

        if (!compact) binding.logList.addItemDecoration(DividerItemDecoration(this, RecyclerView.VERTICAL))

        try {
            logFile = File(applicationContext.filesDir.canonicalPath + "/skyline.log")

            adapter.setItems(logFile.readLines().mapNotNull { logLine ->
                try {
                    val logMeta = logLine.split("|", limit = 3)

                    if (logMeta[0].startsWith("1")) {
                        val level = logMeta[1].toInt()
                        if (level > logLevel) return@mapNotNull null

                        return@mapNotNull LogViewItem(compact, "(" + logMeta[2] + ") " + logMeta[3].replace('\\', '\n'), logLevels[level])
                    } else {
                        return@mapNotNull HeaderViewItem(logMeta[1])
                    }
                } catch (ignored : IndexOutOfBoundsException) {
                } catch (ignored : NumberFormatException) {
                }
                null
            })
        } catch (e : FileNotFoundException) {
            Log.w("Logger", "IO Error during access of log file: " + e.message)
            Toast.makeText(applicationContext, getString(R.string.file_missing), Toast.LENGTH_LONG).show()

            finish()
        } catch (e : IOException) {
            Log.w("Logger", "IO Error during access of log file: " + e.message)
            Toast.makeText(applicationContext, getString(R.string.error) + ": ${e.localizedMessage}", Toast.LENGTH_LONG).show()
        }
    }

    /**
     * This inflates the layout for the menu [R.menu.toolbar_log] and sets up searching the logs
     */
    override fun onCreateOptionsMenu(menu : Menu) : Boolean {
        menuInflater.inflate(R.menu.toolbar_log, menu)

        val searchView = menu.findItem(R.id.action_search_log).actionView as SearchView
        searchView.isSubmitButtonEnabled = false

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query : String) : Boolean {
                searchView.isIconified = false
                return false
            }

            override fun onQueryTextChange(newText : String) : Boolean {
                adapter.filter.filter(newText)
                return true
            }
        })

        return super.onCreateOptionsMenu(menu)
    }

    /**
     * This handles menu selection for [R.id.action_clear] and [R.id.action_share_log]
     */
    override fun onOptionsItemSelected(item : MenuItem) : Boolean {
        return when (item.itemId) {
            R.id.action_clear -> {
                try {
                    logFile.writeText("")
                } catch (e : IOException) {
                    Log.w("Logger", "IO Error while clearing the log file: " + e.message)
                    Toast.makeText(applicationContext, getString(R.string.error) + ": ${e.localizedMessage}", Toast.LENGTH_LONG).show()
                }

                Toast.makeText(applicationContext, getString(R.string.cleared), Toast.LENGTH_LONG).show()
                finish()
                true
            }

            R.id.action_share_log -> {
                uploadAndShareLog()
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    /**
     * This uploads the logs and launches the [Intent.ACTION_SEND] intent
     */
    private fun uploadAndShareLog() {
        Snackbar.make(findViewById(android.R.id.content), getString(R.string.upload_logs), Snackbar.LENGTH_SHORT).show()

        val shareThread = Thread {
            var urlConnection : HttpsURLConnection? = null

            try {
                val url = URL("https://hastebin.com/documents")

                urlConnection = url.openConnection() as HttpsURLConnection
                urlConnection.requestMethod = "POST"
                urlConnection.setRequestProperty("Host", "hastebin.com")
                urlConnection.setRequestProperty("Content-Type", "application/json; charset=utf-8")
                urlConnection.setRequestProperty("Referer", "https://hastebin.com/")

                urlConnection.outputStream.bufferedWriter().use {
                    it.write(logFile.readText())
                    it.flush()
                }

                if (urlConnection.responseCode != 200) {
                    Log.e("LogUpload", "HTTPS Status Code: " + urlConnection.responseCode)
                    throw Exception()
                }

                val bufferedReader = urlConnection.inputStream.bufferedReader()
                val key = JSONObject(bufferedReader.readText()).getString("key")
                bufferedReader.close()

                val result = "https://hastebin.com/$key"
                val sharingIntent = Intent(Intent.ACTION_SEND).setType("text/plain").putExtra(Intent.EXTRA_TEXT, result)

                startActivity(Intent.createChooser(sharingIntent, "Share log url with:"))
            } catch (e : Exception) {
                runOnUiThread { Snackbar.make(findViewById(android.R.id.content), getString(R.string.error) + ": ${e.localizedMessage}", Snackbar.LENGTH_LONG).show() }
                e.printStackTrace()
            } finally {
                urlConnection!!.disconnect()
            }
        }

        shareThread.start()
    }

    /**
     * This handles on calling [onBackPressed] when [KeyEvent.KEYCODE_BUTTON_B] is lifted
     */
    override fun onKeyUp(keyCode : Int, event : KeyEvent?) : Boolean {
        if (keyCode == KeyEvent.KEYCODE_BUTTON_B) {
            onBackPressed()
            return true
        }

        return super.onKeyUp(keyCode, event)
    }
}
